File: apis/src/test/java/com/beat/apis/booking/GuestBookingServiceConcurrencyTest.java
package com.beat.apis.booking;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.transaction.annotation.Transactional;

import com.beat.application.frontoffice.booking.command.GuestBookingCommandService;
import com.beat.application.frontoffice.booking.command.GuestBookingCommand;
import com.beat.application.frontoffice.booking.result.BookingCreationResult;
import com.beat.apis.support.AbstractIntegrationTest;
import com.beat.domain.booking.repository.BookingRepository;
import com.beat.domain.performance.repository.PerformanceRepository;
import com.beat.domain.sharedkernel.vo.BankName;
import com.beat.domain.performance.model.Genre;
import com.beat.domain.performance.model.Performance;
import com.beat.domain.performance.vo.PaymentAccount;
import com.beat.domain.performance.vo.PerformancePeriod;
import com.beat.domain.performance.vo.RunningTime;
import com.beat.domain.performance.vo.TicketPrice;
import com.beat.domain.schedule.repository.ScheduleRepository;
import com.beat.domain.schedule.model.Schedule;
import com.beat.domain.schedule.model.ScheduleNumber;
import com.beat.domain.exception.DomainException;
import com.beat.domain.schedule.exception.ScheduleErrorCode;
import com.beat.domain.user.model.Users;
import com.beat.domain.user.repository.UserRepository;

class GuestBookingServiceConcurrencyTest extends AbstractIntegrationTest {

	private static final Logger logger = LoggerFactory.getLogger(GuestBookingServiceConcurrencyTest.class);
	private static final int CONCURRENT_REQUEST_COUNT = 30;

	@Autowired
	private GuestBookingCommandService guestBookingService;

	@Autowired
	private ScheduleRepository scheduleRepository;

	@Autowired
	private PerformanceRepository performanceRepository;

	@Autowired
	private BookingRepository bookingRepository;

	@Autowired
	private UserRepository userRepository;

	private Schedule schedule1;
	private Schedule schedule2;

	@BeforeEach
	@Transactional
	void setup() {
		logger.info("Setting up initial data...");

		Users maker = createMakerUser();
		Performance performance = createPerformance(maker);
		schedule1 = createSchedule(performance, ScheduleNumber.FIRST, 10);
		schedule2 = createSchedule(performance, ScheduleNumber.SECOND, 1);

		logger.info("Setup completed.");
	}

	@Test
	void testConcurrentGuestBooking() {
		long firstScheduleSuccessCount = executeConcurrentGuestBookings(schedule1, 2, ScheduleNumber.FIRST);
		long secondScheduleSuccessCount = executeConcurrentGuestBookings(schedule2, 1, ScheduleNumber.SECOND);

		assertEquals(5L, firstScheduleSuccessCount);
		assertEquals(1L, secondScheduleSuccessCount);
		assertFinalState();
	}

	private String generateRandomPassword() {
		int randomNum = ThreadLocalRandom.current().nextInt(1000, 10000);
		return String.format("%04d", randomNum);
	}

	private Users createMakerUser() {
		Users maker = Users.create();
		return userRepository.save(maker);
	}

	private Performance createPerformance(Users maker) {
		Performance performance = Performance.create(
			"Performance Title",
			Genre.BAND,
			RunningTime.of(120),
			"Performance Description",
			"Performance Attention Note",
			PaymentAccount.of(BankName.BUSAN, "2342-234234-2344", "이동훈"),
			"poster.jpg",
			"Performance Team",
			"Performance Venue",
			"도로명 주소",
			"상세 주소",
			"123.1111",
			"12.1234",
			"010-1111-1111",
			PerformancePeriod.of(LocalDate.of(2024, 1, 1), LocalDate.of(2024, 12, 31)),
			TicketPrice.of(10000),
			30,
			maker.getId()
		);
		return performanceRepository.save(performance);
	}

	private Schedule createSchedule(Performance performance, ScheduleNumber scheduleNumber, int remainingTicketCount) {
		LocalDateTime performanceDate = LocalDateTime.now().plusDays(1);
		Schedule schedule = Schedule.create(
			performanceDate,
			performanceDate.plusMinutes(performance.getRunningTime()),
			remainingTicketCount,
			scheduleNumber,
			performance.getId()
		);
		return scheduleRepository.save(schedule);
	}

	private long executeConcurrentGuestBookings(
		Schedule schedule,
		int purchaseTicketCount,
		ScheduleNumber scheduleNumber
	) {
		ExecutorService executor = Executors.newFixedThreadPool(CONCURRENT_REQUEST_COUNT);
		CountDownLatch ready = new CountDownLatch(CONCURRENT_REQUEST_COUNT);
		CountDownLatch start = new CountDownLatch(1);
		List<Future<Boolean>> futures = new ArrayList<>();

		for (int i = 0; i < CONCURRENT_REQUEST_COUNT; i++) {
			futures.add(executor.submit(() -> {
				ready.countDown();
				start.await();
				return createGuestBooking(schedule, purchaseTicketCount, scheduleNumber);
			}));
		}

		try {
			assertTrue(ready.await(10, TimeUnit.SECONDS), "Concurrent booking tasks did not become ready");
		} catch (InterruptedException e) {
			executor.shutdownNow();
			Thread.currentThread().interrupt();
			throw new AssertionError("Concurrent booking task setup interrupted", e);
		}
		start.countDown();
		return awaitExecutors(futures, executor);
	}

	private boolean createGuestBooking(Schedule schedule, int purchaseTicketCount, ScheduleNumber scheduleNumber) {
		try {
			BookingCreationResult response =
				guestBookingService.createGuestBooking(
					createGuestBookingRequest(schedule, purchaseTicketCount, scheduleNumber));
			assertNotNull(response);
			return true;
		} catch (DomainException e) {
			if (e.getErrorCode() == ScheduleErrorCode.INSUFFICIENT_TICKETS) {
				return false;
			}
			throw e;
		}
	}

	private GuestBookingCommand createGuestBookingRequest(
		Schedule schedule,
		int purchaseTicketCount,
		ScheduleNumber scheduleNumber
	) {
		return GuestBookingCommand.of(
			schedule.getId(),
			purchaseTicketCount,
			"서지우",
			"010-2222-7196",
			"900101",
			generateRandomPassword()
		);
	}

	private long awaitExecutors(List<Future<Boolean>> futures, ExecutorService executor) {
		executor.shutdown();

		try {
			if (!executor.awaitTermination(120, TimeUnit.SECONDS)) {
				executor.shutdownNow();
			}
		} catch (InterruptedException e) {
			executor.shutdownNow();
			Thread.currentThread().interrupt();
		}

		long successCount = 0L;
		for (Future<Boolean> future : futures) {
			try {
				if (future.get(10, TimeUnit.SECONDS)) {
					successCount++;
				}
			} catch (TimeoutException e) {
				future.cancel(true);
				throw new AssertionError("Concurrent booking task timed out", e);
			} catch (InterruptedException e) {
				future.cancel(true);
				Thread.currentThread().interrupt();
				throw new AssertionError("Concurrent booking task interrupted", e);
			} catch (Exception e) {
				throw new AssertionError("Concurrent booking task failed", e);
			}
		}
		return successCount;
	}

	private void assertFinalState() {
		Schedule firstSchedule = scheduleRepository.findById(schedule1.getId()).orElseThrow();
		Schedule secondSchedule = scheduleRepository.findById(schedule2.getId()).orElseThrow();

		assertEquals(10, firstSchedule.getAllocatedTicketCount());
		assertEquals(1, secondSchedule.getAllocatedTicketCount());

		long firstScheduleBookingCount = bookingRepository.findAll().stream()
			.filter(booking -> Objects.equals(booking.getScheduleId(), firstSchedule.getId()))
			.count();
		long secondScheduleBookingCount = bookingRepository.findAll().stream()
			.filter(booking -> Objects.equals(booking.getScheduleId(), secondSchedule.getId()))
			.count();

		assertEquals(5L, firstScheduleBookingCount);
		assertEquals(1L, secondScheduleBookingCount);
	}
}

File: apis/src/test/kotlin/com/beat/apis/booking/GuestBookingServiceConcurrencyTest.kt
package com.beat.apis.booking

import com.beat.apis.support.BeatTestContainersConfig
import com.beat.application.frontoffice.booking.booker.command.GuestBookingCommand
import com.beat.application.frontoffice.booking.booker.command.GuestBookingCommandService
import com.beat.application.frontoffice.booking.booker.result.BookingCreationResult
import com.beat.application.frontoffice.exception.FrontofficeApplicationErrorType
import com.beat.application.frontoffice.exception.FrontofficeApplicationException
import com.beat.domain.booking.repository.BookingRepository
import com.beat.domain.performance.model.Genre
import com.beat.domain.performance.model.Performance
import com.beat.domain.performance.repository.PerformanceRepository
import com.beat.domain.performance.vo.PaymentAccount
import com.beat.domain.performance.vo.PerformancePeriod
import com.beat.domain.performance.vo.RunningTime
import com.beat.domain.performance.vo.TicketPrice
import com.beat.domain.schedule.exception.ScheduleErrorCode
import com.beat.domain.schedule.model.Schedule
import com.beat.domain.schedule.model.ScheduleNumber
import com.beat.domain.schedule.repository.ScheduleRepository
import com.beat.domain.sharedkernel.vo.BankName
import com.beat.domain.user.model.Users
import com.beat.domain.user.repository.UserRepository
import io.kotest.core.annotation.Tags
import io.kotest.core.spec.IsolationMode
import io.kotest.core.spec.style.FunSpec
import io.kotest.extensions.spring.SpringExtension
import io.kotest.extensions.spring.SpringTestLifecycleMode
import io.kotest.matchers.shouldBe
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Import
import org.springframework.test.context.ActiveProfiles
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.ArrayList
import java.util.Objects
import java.util.concurrent.CountDownLatch
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

@SpringBootTest
@ActiveProfiles("test")
@Import(BeatTestContainersConfig::class)
@Tags("integration", "correctness")
open class GuestBookingServiceConcurrencyTest : FunSpec() {

    @Autowired
    private lateinit var guestBookingService: GuestBookingCommandService

    @Autowired
    private lateinit var scheduleRepository: ScheduleRepository

    @Autowired
    private lateinit var performanceRepository: PerformanceRepository

    @Autowired
    private lateinit var bookingRepository: BookingRepository

    @Autowired
    private lateinit var userRepository: UserRepository

    private lateinit var schedule1: Schedule
    private lateinit var schedule2: Schedule

    init {
        isolationMode = IsolationMode.SingleInstance
        extension(SpringExtension(SpringTestLifecycleMode.Test))

        beforeTest {
            setup()
        }

        test("concurrent guest bookings do not oversell schedules") {
            executeConcurrentGuestBookings(schedule1, 2, ScheduleNumber.FIRST) shouldBe 5L
            executeConcurrentGuestBookings(schedule2, 1, ScheduleNumber.SECOND) shouldBe 1L
            assertFinalState()
        }
    }

    private fun setup() {
        val maker = userRepository.save(Users.create())
        val performance = performanceRepository.save(
            Performance.create(
                performanceTitle = "Performance Title",
                genre = Genre.BAND,
                runningTime = RunningTime.of(120),
                performanceDescription = "Performance Description",
                performanceAttentionNote = "Performance Attention Note",
                paymentAccount = PaymentAccount.of(BankName.BUSAN, "2342-234234-2344", "이동훈"),
                posterImage = "poster.jpg",
                performanceTeamName = "Performance Team",
                performanceVenue = "Performance Venue",
                roadAddressName = "도로명 주소",
                placeDetailAddress = "상세 주소",
                latitude = "123.1111",
                longitude = "12.1234",
                performanceContact = "010-1111-1111",
                performancePeriod = PerformancePeriod.of(
                    LocalDate.of(2024, 1, 1),
                    LocalDate.of(2024, 12, 31),
                ),
                ticketPrice = TicketPrice.of(10_000),
                totalScheduleCount = 30,
                userId = requireNotNull(maker.id),
            ),
        )
        schedule1 = createSchedule(performance, ScheduleNumber.FIRST, 10)
        schedule2 = createSchedule(performance, ScheduleNumber.SECOND, 1)
    }

    private fun createSchedule(
        performance: Performance,
        scheduleNumber: ScheduleNumber,
        remainingTicketCount: Int,
    ): Schedule {
        val performanceDate = LocalDateTime.now().plusDays(1)
        return scheduleRepository.save(
            Schedule.create(
                performanceDate = performanceDate,
                bookingCloseAt = performanceDate.plusMinutes(performance.runningTime.toLong()),
                totalTicketCount = remainingTicketCount,
                scheduleNumber = scheduleNumber,
                performanceId = requireNotNull(performance.id),
            ),
        )
    }

    private fun executeConcurrentGuestBookings(
        schedule: Schedule,
        purchaseTicketCount: Int,
        scheduleNumber: ScheduleNumber,
    ): Long {
        val executor = Executors.newFixedThreadPool(CONCURRENT_REQUEST_COUNT)
        val ready = CountDownLatch(CONCURRENT_REQUEST_COUNT)
        val start = CountDownLatch(1)
        val futures = ArrayList<Future<Boolean>>(CONCURRENT_REQUEST_COUNT)

        repeat(CONCURRENT_REQUEST_COUNT) { requestIndex ->
            futures += executor.submit<Boolean> {
                ready.countDown()
                check(start.await(TASK_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
                    "Concurrent booking tasks did not start"
                }
                createGuestBooking(schedule, purchaseTicketCount, scheduleNumber, requestIndex)
            }
        }

        try {
            if (!ready.await(READY_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
                throw AssertionError("Concurrent booking tasks did not become ready")
            }
        } catch (e: InterruptedException) {
            executor.shutdownNow()
            Thread.currentThread().interrupt()
            throw AssertionError("Concurrent booking task setup interrupted", e)
        }
        start.countDown()
        return awaitExecutors(futures, executor)
    }

    private fun createGuestBooking(
        schedule: Schedule,
        purchaseTicketCount: Int,
        scheduleNumber: ScheduleNumber,
        requestIndex: Int,
    ): Boolean {
        return try {
            val response: BookingCreationResult = guestBookingService.createGuestBooking(
                createGuestBookingRequest(schedule, purchaseTicketCount, scheduleNumber, requestIndex),
            )
            checkNotNull(response)
            true
        } catch (e: FrontofficeApplicationException) {
            if (e.errorCode.code == ScheduleErrorCode.INSUFFICIENT_TICKETS.code &&
                e.errorCode.type == FrontofficeApplicationErrorType.INVALID_INPUT
            ) {
                false
            } else {
                throw e
            }
        }
    }

    private fun createGuestBookingRequest(
        schedule: Schedule,
        purchaseTicketCount: Int,
        scheduleNumber: ScheduleNumber,
        requestIndex: Int,
    ): GuestBookingCommand = GuestBookingCommand.of(
        requireNotNull(schedule.id),
        purchaseTicketCount,
        "서지우",
        "010-%04d-%04d".format(1000 + scheduleNumber.ordinal, 1000 + requestIndex),
        "900101",
        "1234",
    )

    private fun awaitExecutors(
        futures: List<Future<Boolean>>,
        executor: ExecutorService,
    ): Long {
        executor.shutdown()

        try {
            if (!executor.awaitTermination(EXECUTOR_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
                executor.shutdownNow()
            }
        } catch (e: InterruptedException) {
            executor.shutdownNow()
            Thread.currentThread().interrupt()
        }

        var successCount = 0L
        futures.forEach { future ->
            try {
                if (future.get(TASK_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
                    successCount++
                }
            } catch (e: TimeoutException) {
                future.cancel(true)
                throw AssertionError("Concurrent booking task timed out", e)
            } catch (e: InterruptedException) {
                future.cancel(true)
                Thread.currentThread().interrupt()
                throw AssertionError("Concurrent booking task interrupted", e)
            } catch (e: Exception) {
                throw AssertionError("Concurrent booking task failed", e)
            }
        }
        return successCount
    }

    private fun assertFinalState() {
        val firstSchedule = checkNotNull(scheduleRepository.findById(requireNotNull(schedule1.id)))
        val secondSchedule = checkNotNull(scheduleRepository.findById(requireNotNull(schedule2.id)))

        firstSchedule.allocatedTicketCount shouldBe 10
        secondSchedule.allocatedTicketCount shouldBe 1

        val firstScheduleBookingCount = bookingRepository.findAll().count {
            Objects.equals(it.scheduleId, firstSchedule.id)
        }
        val secondScheduleBookingCount = bookingRepository.findAll().count {
            Objects.equals(it.scheduleId, secondSchedule.id)
        }

        firstScheduleBookingCount shouldBe 5
        secondScheduleBookingCount shouldBe 1
    }

    private companion object {
        const val CONCURRENT_REQUEST_COUNT = 30
        const val READY_TIMEOUT_SECONDS = 10L
        const val TASK_TIMEOUT_SECONDS = 10L
        const val EXECUTOR_TIMEOUT_SECONDS = 120L
    }
}
